### PR TITLE
Add even more inductive proofs

### PR DESCRIPTION
**Pull request by h313**
We can now show inductively the `PageAllocator` always increments pointers correctly. Speaking of pointers, the stack pointer arithmetic is now also verified in the same way in `interpreter.rs`, and cleaned up the `Rc.rs` tests and made them inductive.

### CODE REVIEW

File: src/util/paging.rs [hunk 1]
```diff
@@ -229,8 +229,6 @@ impl Page {
         // Make sure out buffer can fit in here
         let final_offset = (aligned_start - self.ptr as usize) + layout.size();
         if final_offset <= self.size {
-            assert!(!self.has_deallocated);
-
```
Comment (Kronos3) [inline src/util/paging.rs:233]: I'm not sure how you can prove this won't trigger? This is a check that blocks the user from doing certain life-cycle level things with the spacewasm engine. It can easily be triggered if say a user tries to load a module into a store when another module failed to load and _didn't_ clear the backing memory.

I think this assert keeps coming up so we should probably add some docs here that describe why it's here and when/what to look for when it triggers

Comment (h313) [reply src/util/paging.rs:233]: Ah oops, that was a dirty edit left over from a previous test. Had to rework the tests in light of this though

**Comment by h313** [comment]
Reworked the invariant testing and caught another edge case where it is possible that pages with a de-alloc may get a new alloc. Added `test_alloc_skips_page_when_teardown` for this situation as well:

https://github.com/nasa/spacewasm/blob/b3f228505b34155f17b84ad2e12c42eb784541de/src/util/paging.rs#L398-L424

File: src/util/paging.rs [hunk 2]
```diff
@@ -63,6 +63,7 @@ struct PageAllocatorInner<A: Allocator, const MAX_PAGES: usize> {
     page_allocator: A,
     page_size: usize,
     pages: [Option<Page>; MAX_PAGES],
+    poisoned: bool,
```
Comment (Kronos3) [inline src/util/paging.rs]: If we do a global poison flag, do we still need a per-page flag?

Comment (h313) [reply src/util/paging.rs]: I kept it in there assuming that other stuff might use the per-page flag. Since I couldn't find any case of that, I've removed it and reworked the proofs to not use it as well.

File: src/util/paging.rs [hunk 3]
```diff
@@ -91,12 +93,16 @@ impl<A: Allocator, const MAX_PAGES: usize> PageAllocatorInner<A, MAX_PAGES> {
             return Err(AllocError::AllocationFailed);
         }
 
+        // All allocations fail if the heap is poisoned
+        if self.poisoned {
+            return Err(AllocError::OutOfMemory);
```
Comment (Kronos3) [inline src/util/paging.rs]: Let's make a new alloc error (`HeapPoisoned`) for this case so the failure is clear.
